src/red_black_tree.jl

```diff
@@ -0,0 +1,327 @@
+    node = search_node(tree, d)
+    return (node.data == d)
+end
+
+function insert_node!(tree::RBTree, node::RBTreeNode)
+    node_y = nothing
+    node_x = tree.root
+
+    while node_x != tree.Nil
+        node_y = node_x
+        if node.data < node_x.data
+            node_x = node_x.leftChild
+        else
+            node_x = node_x.rightChild
+        end
+    end
+
+    node.parent = node_y
+    if node_y == nothing
+        tree.root = node
+    elseif node.data < node_y.data
+        node_y.leftChild = node
+    else
+        node_y.rightChild = node
+    end
+end
+
+function left_rotate!(tree::RBTree, node_x::RBTreeNode)
+    node_y = node_x.rightChild
+    node_x.rightChild = node_y.leftChild
+    if node_y.leftChild != tree.Nil
+        node_y.leftChild.parent = node_x
+    end
+    node_y.parent = node_x.parent
+    if (node_x.parent == nothing)
+        tree.root = node_y
+    elseif (node_x == node_x.parent.leftChild)
+        node_x.parent.leftChild = node_y
+    else
+        node_x.parent.rightChild = node_y
+    end
+    node_y.leftChild = node_x
+    node_x.parent = node_y
+end    
+
+function right_rotate!(tree::RBTree, node_x::RBTreeNode)
+    node_y = node_x.leftChild
+    node_x.leftChild = node_y.rightChild
+    if node_y.rightChild != tree.Nil
+        node_y.rightChild.parent = node_x
+    end
+    node_y.parent = node_x.parent
+    if (node_x.parent == nothing)
+        tree.root = node_y
+    elseif (node_x == node_x.parent.leftChild)
+        node_x.parent.leftChild = node_y
+    else
+        node_x.parent.rightChild = node_y
+    end
+    node_y.rightChild = node_x
+    node_x.parent = node_y
+end    
+
+function fix_insert!(tree::RBTree, node::RBTreeNode)
+    parent = nothing
+    grand_parent = nothing
+    # for root node, we need to change the color to black
+    # other nodes, we need to maintain the property such that
+    # no two adjacent nodes are red in color
+    while  node != tree.root && node.parent.color
+        parent = node.parent
+        grand_parent = parent.parent
+        
+        # parent is the leftChild of grand_parent
+        if (parent == grand_parent.leftChild)
+            uncle = grand_parent.rightChild
+
+            # uncle is red in color
+            if (uncle.color)
+                grand_parent.color = true
+                parent.color = false
+                uncle.color = false
+                node = grand_parent
+            # uncle is black in color
+            else 
+                # node is rightChild of it's parent
+                if (node == parent.rightChild)
+                    node = parent
+                    left_rotate!(tree, node)
+                end
+                # node is leftChild of it's parent
+                node.parent.color = false
+                node.parent.parent.color = true
+                right_rotate!(tree, node.parent.parent)
+            end
+        # parent is the rightChild of grand_parent
+        else 
```

Comment:
I am unclear as to if this comment belongs to the block following the else or the block preciding it.
(Similar for eom other comments in this section)
If it is the block following then it can be clearer if put on same line.
If not then it should be indented
```suggestion
        else  # parent is the rightChild of grand_parent
```